control-plane/csi-driver/src/bin/node/nodeplugin_svc.rs

```diff
@@ -177,3 +104,67 @@ pub(crate) async fn find_mount(
         Ok(None)
     }
 }
+
+/// /// Issue FIFREEZE and FITHAW ioctl on the mount_path derived from the volume_id based on the
+/// command.
+pub(crate) async fn fsfreeze(volume_id: &str, command: FsFreezeOpt) -> Result<(), ServiceError> {
+    let uuid = Uuid::parse_str(volume_id).context(InvalidVolumeId {
+        volume_id: volume_id.to_string(),
+    })?;
+
+    if let Some(device) = Device::lookup(&uuid).await.context(InternalFailure {
+        volume_id: volume_id.to_string(),
+    })? {
+        let device_path = device.devname();
+        if let Some(mnt) = mount::find_mount(Some(&device_path), None) {
+            // Make a preflight check, to ensure subsystem has atleast one live path.
+            let device_nqn = device.devnqn();
+            fsfreeze_preflight_check(volume_id, &device_nqn).map_err(|errno| {
+                ServiceError::FsfreezeFailed {
+                    volume_id: volume_id.to_string(),
+                    errno,
+                }
+            })?;
+
+            // Execute the fsfreeze call.
+            let dest = mnt.dest.display().to_string();
+            return fsfreeze_ioctl(command, dest).await.map_err(|errno| {
+                ServiceError::FsfreezeFailed {
+                    volume_id: volume_id.to_string(),
+                    errno,
+                }
+            });
+        } else {
+            // mount::find_mount does not return any matches,
+            // for mounts which are bind mounts to block devices
+            // (raw block volume).
+            // It would be incorrect to return the VolumeNotFound error,
+            // if the volume is mounted as a raw block volume on this node.
+            // Use findmnt to work out if volume is mounted as a raw
+            // block, i.e. we get some matches, and return the
+            // BlockDeviceMount error.
+            let mountpaths = findmnt::get_mountpaths(&device_path).context(InternalFailure {
+                volume_id: volume_id.to_string(),
+            })?;
+            if !mountpaths.is_empty() {
+                debug!(
+                    "{} for volume_id :{} : failed for block device",
+                    command, volume_id
+                );
+                return Err(ServiceError::BlockDeviceMount {
+                    volume_id: volume_id.to_string(),
+                });
+            }
+            debug!(
+                "{} for volume_id :{} : failed, cannot find volume",
+                command, volume_id
+            );
+        }
+        return Err(ServiceError::VolumeNotMounted {
+            volume_id: volume_id.to_string(),
+        });
```

Comment:
```suggestion
        Err(ServiceError::VolumeNotMounted {
            volume_id: volume_id.to_string(),
        })
```

Reply:
This is needed ?

Reply:
Right, done. I did not change the inner `if let` as that would first execute the non filesystem bit first.